### PR TITLE
feat: add fluentbit monitoring

### PR DESCRIPTION
**Pull request by gnomex909**
This adds FluentBit monitoring. FluentBit is default technology used by Firelens, which is suggested log router to use with ECS. Current implementation doesn't support any alarms, but can be extended if there will be interest. 

Photo of this dashboard segment from our team code that this change is based upon.
screenshot_2023-12-12_at_15 48 19

---

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

### CODE REVIEW

File: lib/monitoring/fluentbit/FluentBitMetricFactory.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+  constructor(scope: MonitoringScope, props: FluentBitMetricFactoryProps) {
+    this.scope = scope;
+    this.metricFactory = scope.createMetricFactory();
+    this.namespace =
+      props.namespace ??
+      this.metricFactory.getNamespaceWithFallback(props.namespace);
+  }
+
+  fluentBitFilterMetrics(logGroup: ILogGroup) {
```
Comment (echeung-amzn) [inline lib/monitoring/fluentbit/FluentBitMetricFactory.ts:33]: The `fluentBit` prefixes seem redundant.

File: lib/monitoring/fluentbit/FluentBitMonitoring.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    this.fluentBitOutputMetrics = this.metricFactory.fluentBitOutputMetrics(
+      props.logGroup
+    );
+    this.fluentBitFilterMetrics = this.metricFactory.fluentBitFilterMetrics(
+      props.logGroup
+    );
+    this.metricFactory.fluentBitMetricsWithoutWidgets(props.logGroup);
+  }
+
+  widgets(): IWidget[] {
```
Comment (echeung-amzn) [inline lib/monitoring/fluentbit/FluentBitMonitoring.ts:56]: Might want to have something for `summaryWidgets` too?

File: lib/monitoring/fluentbit/FluentBitMonitoring.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  private createFluentBitMetricWidget(
+    metrics: MetricWithAlarmSupport[],
+    title: string
+  ): GraphWidget {
+    return new GraphWidget({
+      width: HalfWidth,
+      height: DefaultGraphWidgetHeight,
+      title: `${title}`,
```
Comment (echeung-amzn) [inline lib/monitoring/fluentbit/FluentBitMonitoring.ts:102]:
```suggestion
      title,
```

**Review by echeung-amzn — CHANGES_REQUESTED** [review]

File: lib/monitoring/fluentbit/FluentBitMetricFactory.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      );
+      const metric = metricFilter.metric({
+        statistic: MetricStatistic.MAX,
+      });
+      storageMetrics.push(metric);
+    });
+    return storageMetrics;
+  }
+
+  fluentBitMetricsWithoutWidgets(logGroup: ILogGroup) {
```
[the remark this answers is not in the document]
Comment (gnomex909) [reply lib/monitoring/fluentbit/FluentBitMetricFactory.ts:100]: It creates Metric Filters for metrics that we do not have on dashboard. Those are either not useless outside of detailed investigation and/or totally different number scale than other metrics (input bytes total, output bytes totals and output records)

[the remark this answers is not in the document]
Comment (echeung-amzn) [reply lib/monitoring/fluentbit/FluentBitMetricFactory.ts]: Can we make that clear in the API (e.g, maybe with a prop that creates these extra metric filters)? That way users will be aware of such useful metric filters, and also won't be surprised that they're paying for some metric filters they never end up using.

[the remark this answers is not in the document]
Comment (gnomex909) [reply lib/monitoring/fluentbit/FluentBitMetricFactory.ts]: To understand better your comment - you suggest to add additional property to Monitoring class that will decide if we are creating metric filters for metrics not shown on default graphs (and if we call this function at all)? It's cool suggestion, but I want to be sure that we are on the same page

[the remark this answers is not in the document]
Comment (echeung-amzn) [reply lib/monitoring/fluentbit/FluentBitMetricFactory.ts]: Yes, a prop in `FluentBitMonitoringProps`.

[the remark this answers is not in the document]
Comment (gnomex909) [reply lib/monitoring/fluentbit/FluentBitMetricFactory.ts]: It was added in latest commit

**Review by echeung-amzn — APPROVED** [review]

**Pull request merged**